Sprint 16

lmazulova opened the pull request.

MazulovaLisa committed:
feat: Добавлен экран онбординга

MazulovaLisa committed:
feat: Добавлена возможность добавления новой категории.

Нужно переписать на MVVM, и исправить мелкие баги.

MazulovaLisa committed:
feat: Первая версии на проверку

Исправлены все баги, завершен перенос category на MVVM

DanilaG commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,87 @@
+    }()
+    
+    lazy var pages: [UIViewController] = {
+        return [OnboardingPage(state: .firstScreen), OnboardingPage(state: .secondScreen)]
```
Использование enum для обозначения экрана онбординга может привести к следующим не желательным явлениям:
- Росту enum-а + трудности в читаймости, так как сторонним разработчкиу не понятно что за `firstScreen` или `anotherScreen`
- При добавлении экрана придётся дорабатывать `OnboardingPage`, что приводит к рискам сломать остальные экраны 
- Росту сложности класса `OnboardingPage` и количеству if-ов в них
- Появлению различных костылей, если `OnboardingPage` не будет отображать все доступные типы экранов

Что можно сделать:
- В инит `OnboardingPage` передавать сразу нужные данные для отображения (картинка фона, текста)

DanilaG commented on Tracker/Onboarding/OnboardingPage.swift:
```diff
@@ -0,0 +1,90 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    var state: OnboardingPageState
```
Давай сделаем всё что можно приватным

DanilaG commented on Tracker/Trackers/AddingNewCategory/Helpers/CategoryCell.swift:
```diff
@@ -0,0 +1,52 @@
+        ])
+    }
+    
+    func setup(with model: CategoryCellViewModel) {
```
Обычно, не приватные функции размещают над приватными

DanilaG commented on Tracker/Trackers/AddingNewCategory/ViewModels/СategoryViewModel.swift:
```diff
@@ -0,0 +1,91 @@
+
+typealias Binding<T> = (T) -> Void
+
+protocol CategoryViewModelProtocol {
```
Давай добавим `AnyObject`. Это помогает ускорять программу + снижает шансы допустить утечки памяти

DanilaG commented on Tracker/Trackers/AddingNewCategory/ViewModels/СategoryViewModel.swift:
```diff
@@ -0,0 +1,91 @@
+protocol CategoryViewModelProtocol {
+    func numberOfRows() -> Int
+    func addRecord(with title: String)
+    func cellViewModel(at indexPath: IndexPath) -> CategoryCellViewModel
+    func selectCategory(at indexPath: IndexPath)
```
Не очень хорошо, когда в названиях методов view model отражаются детали реализации ui. Сегодня оно может быть на ячейках UIKit, завтра это будет стек или ещё что-то. Рекомендую назвать класс списка как `CategoriesListViewModel`, модель отдельной категории `CategoryViewModel`. А методы `CategoriesListViewModel` будут работать не с ячейками, а категориями (`cellViewModel(at indexPath: IndexPath) -> CategoryCellViewModel` станет `сategory(at index: Index) -> CategoryViewModel`, `numberOfRows() -> Int` станет `numberOfCategories() -> Int` и т д)

DanilaG commented on Tracker/Trackers/AddingNewCategory/ViewModels/СategoryViewModel.swift:
```diff
@@ -0,0 +1,91 @@
+    
+    private var selectedIndexPath: IndexPath?
+    
+    var visibleDataChanged: Binding<TrackerStoreUpdate>?
+    
+    var selectedCategoryTitle: String? {
```
Обычно, не приватные переменные объявляются над приватными

MazulovaLisa committed:
fix: Исправления после ревью

Исправлен баг со скрытием клавиатуры, переписан инициализатор onboarding экрана, исправлен баг с выбором категорий с одинаковыми названиями.

MazulovaLisa committed:
fix: Исправлен баг с клавиатурой при создании нового трекера, абстрагировала viewModel от tableView.
